File: le-liaison/src/main/python/liaison/QueryColumn.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

from .exceptions import UnknownMetadataValue

class QueryColumn(object):
 
    def __init__(self, name, expression, approved_usage = None, display_name = None,
                 category = None, statistical_type = None, tags = None,
                 fundamental_type = None, description = None,
                 display_discretization = None):

        self.initFromValues(name, expression, approved_usage, display_name, category,
                            statistical_type, tags, fundamental_type,
                            description, display_discretization)


    def getName(self):
        return self.name_

    def setName(self, n):
        self.name_ = n

    def getExpression(self):
        return self.exp_

    def setExpression(self, e):
        self.exp_ = e

    def getApprovedUsage(self):
        return self.approved_usage_

    def setApprovedUsage(self, au):
        if au is not None and au not in [ 'None', 'Model', 'ModelAndModelInsights', 'ModelAndAllInsights' ]:
            raise UnknownMetadataValue(au)
        self.approved_usage_ = au

    def getDisplayName(self):
        return self.display_name_

    def setDisplayName(self, dn):
        if dn is not None and len(dn) > 46:
            print 'Warning: truncating DisplayName \'{0}\' to 46 characters'.format(dn)
            self.display_name_ = dn[:46]
        else:
            self.display_name_ = dn

    def getCategory(self):
        return self.category_

    def setCategory(self, c):
        if c is not None and len(c) > 22:
            print 'Warning: truncating Category \'{0}\' to 22 characters'.format(c)
            self.category_ = c[:22]
        else:
            self.category_ = c

    def getStatisticalType(self):
        return self.statistical_type_

    def setStatisticalType(self, st):
        if st is not None and st not in [ 'nominal', 'ratio', 'ordinal', 'interval' ]:
            raise UnknownMetadataValue(st)
        self.statistical_type_ = st

    def getTags(self):
        return self.tags_

    def setTags(self, t):
        if t is not None and t.lower() not in [ 'internal', 'external' ]:
            raise UnknownMetadataValue(t)
        self.tags_ = t

    def getFundamentalType(self):
        return self.fundamental_type_

    def setFundamentalType(self, ft):
        if ft is not None and ft not in [ 'alpha', 'numeric', 'currency', 'percentage', 'boolean', 'year' ]:
            raise UnknownMetadataValue(ft)
        self.fundamental_type_ = ft

    def getDescription(self):
        return self.description_

    def setDescription(self, d):
        if d is not None and len(d) > 150:
            print 'Warning: truncating Description \'{0}\' to 150 characters'.format(d)
            self.description_ = d[:150]
        else:
            self.description_ = d

    def getDisplayDiscretization(self):
        return self.display_discretization_

    def setDisplayDiscretization(self, dd):
        self.display_discretization_ = dd

    def setMetadata(self, metadata_type, metadata_value):
        if metadata_type == 'ApprovedUsage':
            self.setApprovedUsage(metadata_value)
        elif metadata_type == 'DisplayName':
            self.setDisplayName(metadata_value)
        elif metadata_type == 'Category':
            self.setCategory(metadata_value)
        elif metadata_type == 'StatisticalType':
            self.setStatisticalType(metadata_value)
        elif metadata_type == 'Tags':
            self.setTags(metadata_value)
        elif metadata_type == 'FundamentalType':
            self.setFundamentalType(metadata_value)
        elif metadata_type == 'Description':
            self.setDescription(metadata_value)
        elif metadata_type == 'DisplayDiscretization':
            self.setDisplayDiscretization(metadata_value)

    def definition(self):
        raise NotImplementedError('QueryColumn.definition()')

    def initFromValues(self, name, expression, approved_usage = None, display_name = None,
                       category = None, statistical_type = None, tags = None,
                       fundamental_type = None, description = None,
                       display_discretization = None):

        self.setName(name)
        self.setExpression(expression)
        self.setApprovedUsage(approved_usage)
        self.setDisplayName(display_name)
        self.setCategory(category)
        self.setStatisticalType(statistical_type)
        self.setTags(tags)
        self.setFundamentalType(fundamental_type)
        self.setDescription(description)
        self.setDisplayDiscretization(display_discretization)

File: le-liaison/src/main/java/com/latticeengines/liaison/exposed/service/QueryColumn.java
package com.latticeengines.liaison.exposed.service;

import com.latticeengines.liaison.exposed.exception.BadMetadataValueException;
import com.latticeengines.liaison.exposed.exception.DefinitionException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public abstract class QueryColumn {

    private String name;
    private String expression;
    private Map<String,String> metadata;
    
    //private static final int maxDisplayNameLength = 64;
    private static final int maxDisplayNameLength = 256;
    private static final int maxCategoryLength = 22;
    private static final int maxDescriptionLength = 150;
    private static final Set<String> approvedUsageValues = new HashSet<String>( Arrays.asList("None","Model","ModelAndModelInsights","ModelAndAllInsights") );
    private static final Set<String> statisticalTypeValues = new HashSet<String>( Arrays.asList("nominal","ratio","ordinal","interval") );
    private static final Set<String> tagsValues = new HashSet<String>( Arrays.asList("Internal","External") );
    private static final Set<String> fundamentalTypeValues = new HashSet<String>( Arrays.asList("unknown","nominal","float","alpha","numeric","currency","percentage","boolean","year") );
    
    
    public QueryColumn( String definition ) throws DefinitionException {
        initFromDefinition( definition );
    }
    
    public QueryColumn( String name, String expression, Map<String,String> metadata ) {
        initFromValues( name, expression, metadata );
    }
    
    public QueryColumn( QueryColumn other ) {
        name = other.name;
        expression = other.expression;
        metadata = new HashMap<>( other.metadata );
    }
    
    public String getName() {
        return name;
    }
    
    public void setName( String name ) {
        this.name = name;
    }
    
    public String getExpression() {
        return expression;
    }
    
    public void setExpression( String expression ) {
        this.expression = expression;
    }
    
    public void setApprovedUsage( String value ) throws BadMetadataValueException {
        if( !approvedUsageValues.contains(value) ) {
            throw new BadMetadataValueException( String.format("Cannot set ApprovedUsage to %s", value) );
        }
        metadata.put( "ApprovedUsage", value );
    }
    
    public void setDisplayName( String value ) throws BadMetadataValueException {
        if( value.length() > maxDisplayNameLength ) {
            throw new BadMetadataValueException( String.format("DisplayName \'%s\' is longer than %d", value, maxDisplayNameLength) );
        }
        metadata.put( "DisplayName", value );
    }
    
    public void setCategory( String value ) throws BadMetadataValueException {
        if( value.length() > maxCategoryLength ) {
            throw new BadMetadataValueException( String.format("Category \'%s\' is longer than %d", value, maxCategoryLength) );
        }
        metadata.put( "Category", value );
    }
    
    public void setStatisticalType( String value ) throws BadMetadataValueException {
        if( !statisticalTypeValues.contains(value.toLowerCase()) ) {
            throw new BadMetadataValueException( String.format("Cannot set StatisticalType to %s", value) );
        }
        metadata.put( "StatisticalType", value.toLowerCase() );
    }
    
    public void setTags( String value ) throws BadMetadataValueException {
        if( !tagsValues.contains(value) ) {
            throw new BadMetadataValueException( String.format("Cannot set Tags to %s", value) );
        }
        metadata.put( "Tags", value );
    }
    
    public void setFundamentalType( String value ) throws BadMetadataValueException {
        if( !fundamentalTypeValues.contains(value.toLowerCase()) ) {
            throw new BadMetadataValueException( String.format("Cannot set FundamentalType to %s", value) );
        }
        if( value.equals("Unknown") ) {
            metadata.put( "FundamentalType", value );
        }
        else {
            metadata.put( "FundamentalType", value.toLowerCase() );
        }
    }
    
    public void setDescription( String value ) throws BadMetadataValueException {
        if( value.length() > maxDescriptionLength ) {
            throw new BadMetadataValueException( String.format("Description \'%s\' is longer than %d", value, maxDescriptionLength) );
        }
        metadata.put( "Description", value );
    }
    
    public void setDisplayDiscretization( String value ) {
        metadata.put( "DisplayDiscretizationStrategy", value );
    }
    
    public void setMetadata( Map<String,String> metadata ) throws BadMetadataValueException {
        for( Map.Entry<String,String> entry : metadata.entrySet() ) {
            setMetadata( entry.getKey(), entry.getValue() );
        }
    }
    
    public void setMetadata( String metadataType, String metadataValue ) throws BadMetadataValueException {
        if( metadataType.equals("ApprovedUsage") ) {
            setApprovedUsage( metadataValue );
        }
        else if ( metadataType.equals("DisplayName") ) {
            setDisplayName( metadataValue );
        }
        else if ( metadataType.equals("Category") ) {
            setCategory( metadataValue );
        }
        else if ( metadataType.equals("StatisticalType") ) {
            setStatisticalType( metadataValue );
        }
        else if ( metadataType.equals("Tags") ) {
            setTags( metadataValue );
        }
        else if ( metadataType.equals("FundamentalType") ) {
            setFundamentalType( metadataValue );
        }
        else if ( metadataType.equals("Description") ) {
            setDescription( metadataValue );
        }
        else if ( metadataType.equals("DisplayDiscretizationStrategy") ) {
            setDisplayDiscretization( metadataValue );
        }
        else {
            throw new BadMetadataValueException( String.format("\"%s\" is not a valid metadata type", metadataType) );
        }
    }
    
    public Map<String,String> getMetadata() {
        return metadata;
    }
    
    public void initFromValues( String name, String expression, Map<String,String> metadata ) {
        this.name = name;
        this.expression = expression;
        this.metadata = new HashMap<>();
        setMetadata( metadata );
        if ( !this.metadata.containsKey("Tags") ) {
            setTags("Internal");
        }
        if ( this.metadata.get("Tags").equals("Internal") && !this.metadata.containsKey("Category") ) {
            setCategory("Lead Information");
        }
    }
    
    public abstract void initFromDefinition( String defn ) throws DefinitionException;
    
    public abstract String definition();
}
